crates/lair_keystore_api/src/internal/crypto_box.rs

```diff
@@ -0,0 +1,179 @@
+#[derive(Debug, PartialEq)]
+pub struct CryptoBoxNonce([u8; NONCE_BYTES]);
+
+impl CryptoBoxNonce {
+    fn new_random() -> Self {
+        let mut rng = rand::thread_rng();
+        let mut bytes = [0; NONCE_BYTES];
+        // We rely on the lib_crypto_box nonce length being the same as what we expect.
+        // Should be a reasonably safe bet as 24 bytes is dictated by the crypto_box algorithm.
+        bytes.copy_from_slice(
+            lib_crypto_box::generate_nonce(&mut rng).as_slice()
+        );
+        Self(bytes)
+    }
+}
+
+impl AsRef<[u8; NONCE_BYTES]> for CryptoBoxNonce {
+    fn as_ref(&self) -> &[u8; NONCE_BYTES] {
+        &self.0
+    }
+}
+
+impl AsRef<[u8]> for CryptoBoxNonce {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl From<[u8; NONCE_BYTES]> for CryptoBoxNonce {
+    fn from(array: [u8; NONCE_BYTES]) -> Self {
+        Self(array)
+    }
+}
+
+impl std::convert::TryFrom<&[u8]> for CryptoBoxNonce {
+    type Error = crate::error::LairError;
+    fn try_from(slice: &[u8]) -> Result<Self, Self::Error> {
+        if slice.len() == NONCE_BYTES {
+            let mut inner = [0; NONCE_BYTES];
+            inner.copy_from_slice(slice);
+            Ok(Self(inner))
+        }
+        else {
+            Err(crate::error::LairError::CryptoBoxNonceLength)
+        }
+    }
+}
+
+impl CryptoBoxNonce {
+    /// Always NONCE_BYTES.
+    pub fn len(&self) -> usize {
+        NONCE_BYTES
+    }
+}
+
+/// "Additional associated data" as per the aead rust crate Payload.
+/// May be empty. Must be valid if present.
+pub struct CryptoBoxAad(Vec<u8>);
+
+/// The nonce and encrypted data together.
+/// @todo include additional associated data?
+#[derive(Debug, PartialEq)]
+pub struct CryptoBoxEncryptedData {
+    /// The nonce generated during encryption.
+    /// We never allow nonce to be set externally so we need to return it.
+    pub nonce: CryptoBoxNonce,
+    /// The encrypted version of our input data.
+    pub encrypted_data: Arc<Vec<u8>>,
+}
+
+/// Data to be encrypted.
+/// Not associated with a nonce because we enforce random nonces.
+#[derive(Debug, PartialEq, Clone)]
+pub struct CryptoBoxData {
+    /// Data to be encrypted.
+    pub data: Arc<Vec<u8>>,
+}
+
+impl AsRef<[u8]> for CryptoBoxData {
+    fn as_ref(&self) -> &[u8] {
+        self.data.as_ref()
+    }
+}
+
+impl CryptoBoxData {
+    /// Length of newtype is length of inner.
+    pub fn len(&self) -> usize {
+        AsRef::<[u8]>::as_ref(self).len()
+    }
+}
+
+impl From<Vec<u8>> for CryptoBoxData {
+    fn from(v: Vec<u8>) -> Self {
+        Self {
+            data: Arc::new(v)
+        }
+    }
+}
+
+/// Wrapper around crypto_box from whatever lib we use.
+/// No BYO nonces. Nonces always random and returned as part of `CryptoBoxEncryptedData`.
+/// No BYO algorithms (cipher agility). Algorithm always X25519XSalsa20Poly1305.
+/// Currently no additional associated data but DNA space may be included in the future.
```

Comment:
This makes it sound like we will never allow flexibility...

I'm totally cool with our first cut providing a minimal api, but eventually I'd like to just expose as much of the libsodium api as we can, and let devs decide how complex they want to make things... we can totally make options with a ::default() that fills in a None for nonce that causes a random nonce to be generated, but let people fill that out if the care for some reason.

Similarly - we don't know that dna is the only associated data people will ever want..

&& eventually, the only api I really want is `crypto_aead_xchacha20poly1305_ietf_*`

To be clear, I'm only asking for a documentation change right now to not make it seem so final.

Reply:
@neonphog i agree 100% not my intent to make this sound set in stone